Automattic/wp-calypso
# Gutenberg: Add the Calypso Media Modal

#### Changes proposed in this Pull Request

* Add the Calypso Media Modal in Gutenberg.
* Wire it up with the `Image`, `Cover Image`, `Gallery`, `File`, and `Inline Image` blocks.

#### Notes

- Gutenberg's `Gallery` block doesn't seem to support the same options as Calypso gallery shortcode (e.g. the layout), but it simply has a slider for the number of columns.
This is why I've decided, for now, to disable the edit gallery view in the media modal.

- For some reasons, when opening the modal through the `Inline Image` block, the `onOpen` function is called _a lot_ of times. The `setState` contained inside is called as many times, to the point that it crashes the modal.
I couldn't figure out why (in Core it is called only once, but then again, the logic is very different), but I've worked around this issue by [only calling `setState` if needed](https://github.com/Automattic/wp-calypso/pull/27687/files#diff-da6effaf543f6ec6542895c04d32ce4cR42).

- Modifying media in the modal automatically updates the blocks using those media by listening to `MediaStore.on( 'change' )`.
Currently though, `Inline Image` blocks are not updated accordingly and, apparently, if their image has been changed, they will be removed on reload.

- `Audio` and `Video` blocks will be handled in one or more follow up PRs.

#### Testing instructions



* Open Gutenberg in Calypso dev: http://calypso.localhost:3000/gutenberg/post/
* Add an `Image` block.
* Click on the Media Library button.
* Make sure the Media Modal shows up, and the images are loaded as expected.
* Try inserting one image and make sure it's indeed added to the block.
* Try reloading the editor and check if the `Image` block shows the added image.
* Try editing the block and check that the image is selected in the Media Library.
* Do the same tests for `Cover Image`, `Gallery`, and `File`.
* Move the cursor inside a Paragraph block and try inserting an `Inline Image`. Make sure the media modal shows up, it's possible to insert an inline image in the paragraph, and that the block still works after a save and reload.

Fixes #26904
Fixes #27641
